### PR TITLE
Fix root volume resizing on EC2 KVM instances (M5, C5, etc) 

### LINKED ISSUE #33092: growpart on new (C5/M5) AWS instances fails to increase root partition
## Issue description

When trying to use the new AWS C5 and M5 instance types `growpart` fails with an error in stage-1-init:
```
[    2.942859] stage-1-init: FAILED: partition-number must be a number
```
While investigating I found that these new instances use a different storage driver for the root device:
http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/nvme-ebs-volumes.html
This causes issue because the way growpart is called in stage-1-init from here:
https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/virtualisation/grow-partition.nix#L36-L44
```
    boot.initrd.postDeviceCommands = ''
      rootDevice="${config.fileSystems."/".device}"
      if [ -e "$rootDevice" ]; then
        rootDevice="$(readlink -f "$rootDevice")"
        parentDevice="$(lsblk -npo PKNAME "$rootDevice")"
        TMPDIR=/run sh $(type -P growpart) "$parentDevice" "''${rootDevice#$parentDevice}"
        udevadm settle
      fi
    '';
```
calls it with `partition-number` of `p1` instead of just `1`: (boot.trace enabled dmesg output)
```
[    3.110620] stage-1-init: + TMPDIR=/run sh /nix/store/wv2hhn6xcnfqgd7blq8fp27gg8fvzxqj-extra-utils/bin/growpart /dev/nvme0n1 p1
[    3.111210] stage-1-init: FAILED: partition-number must be a number
```

I found what I believe is a related issue: https://bugs.launchpad.net/cloud-initramfs-tools/+bug/1697695

So that I can keep working on testing out these new instance types I put a probably really gross hack in a fork of nixpkgs:
https://github.com/extole/nixpkgs/pull/3/files#diff-8e04ff5e80705e77d2a4ff12ea3dda78
```
-        TMPDIR=/run sh $(type -P growpart) "$parentDevice" "''${rootDevice#$parentDevice}"
 +        # support nvme style partitions on AWS
 +        if echo "$parentDevice" | grep nvme >/dev/null 2>&1; then
 +          partitionNumber="''${rootDevice#''${parentDevice}p}"
 +        else
 +          partitionNumber="''${rootDevice#$parentDevice}"
 +        fi
 +        TMPDIR=/run sh $(type -P growpart) "$parentDevice" "$partitionNumber"
```

### Steps to reproduce

Build an AMI off of nixpkgs@67f9fd9e9f1b8e6e97e5f495ebf5e58f79c5c318 and deploy the AMI with a root device definition larger than the AMI so it tries to resize the partition on boot. Login to the instance and note that its root partition is the AMI's size not what you gave it on launch.

## Technical details

Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the
results.
```
 - system: `"x86_64-linux"`
 - host os: `Linux 4.9.70, NixOS, 18.03pre-git (Impala)`
 - multi-user?: `no`
 - sandbox: `yes`
 - version: `nix-env (Nix) 1.11.16`
 - channels(root): `"nixos-18.03pre-git"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos/nixpkgs`
```

### PR DESCRIPTION
**Pull request by ngortheone**
Fixes:
https://github.com/NixOS/nixpkgs/issues/33417
https://github.com/NixOS/nixpkgs/issues/33092

Root volume resizing was broken on new KVM instances because of the way partition name was computed
```
rootDevice="$(readlink -f "$rootDevice")"
        parentDevice="$(lsblk -npo PKNAME "$rootDevice")"
        TMPDIR=/run sh $(type -P growpart) "$parentDevice" "''${rootDevice#$parentDevice}"
```
`"''${rootDevice#$parentDevice}"` results in `p1` instead of `1` when root device is nvme disk

resulting in small root partition
```
# lsblk
NAME        MAJ:MIN RM SIZE RO TYPE MOUNTPOINT
nvme0n1     259:0    0  32G  0 disk
└─nvme0n1p1 259:2    0  2G  0 part /
```

Suggested solution borrowed from here:
https://git.launchpad.net/cloud-initramfs-tools/tree/growroot/scripts/local-bottom/growroot
(lines 38 - 49)

Performed testing on m4 and m5 instances and confirmed that fix works

### CODE REVIEW

**Review by copumpkin — APPROVED** [review]
Looks good to me, thanks!

We should probably also backport this to 18.03 and regenerate the AMIs, @edolstra, right?

**Comment by copumpkin** [comment]
cc @greglearns @polynomial @hakujin who have also expressed interest in the related tickets

**Comment by hakujin** [comment]
@copumpkin @edolstra how can I help get this fix into new AMIs? New to nix but happy to contribute if someone will point me in the correct direction.

**Comment by copumpkin** [comment]
We've discovered one remaining bug that makes the new instance types fail if you use large root EBS volumes with NixOS. Hoping to get a fix to that in today, then whenever @edolstra has time to make the AMIs, I think we'll be set.

**Comment by edolstra** [comment]
New AMIs are up!

**Comment by copumpkin** [comment]
Thanks 👍

**Comment by zeme-wana** [comment]
Thank you @edolstra

**Comment by cristiam86** [comment]
I found the solution [https://stackoverflow.com/a/52563443/637623](here)

TLDR: if your partition name is nvme0n1p1, then your growpart command should be like this
`sudo growpart /dev/nvme0n1 1` (without the letterp)

**Comment by copumpkin** [comment]
@cristiam86 sorry, the solution to what? is this not working for you?

**Comment by cristiam86** [comment]
I mean that I wasn't able to understand the solution given by @ngortheone. So I looked a little bit further and found that StackOverflow post were the problem was explained in a more simple way.

**Comment by copumpkin** [comment]
Oh, if you have a simpler one that still works a PR would be very nice!

**Comment by cristiam86** [comment]
Oh no, sorry my problem wasn't linked with this project was resizing the partition manually, but looking for solutions I stumble across this repo and helped me to reach the final solution.
Thanks
